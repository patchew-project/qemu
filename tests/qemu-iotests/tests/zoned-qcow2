#!/usr/bin/env bash
#
# Test zone management operations for qcow2 file.
#

seq="$(basename $0)"
echo "QA output created by $seq"
status=1 # failure is the default!

file_name="zbc.qcow2"
_cleanup()
{
  _cleanup_test_img
  _rm_test_img "$file_name"
}
trap "_cleanup; exit \$status" 0 1 2 3 15

# get standard environment, filters and checks
. ../common.rc
. ../common.filter
. ../common.qemu

# This test only runs on Linux hosts with qcow2 image files.
_supported_fmt qcow2
_supported_proto file
_supported_os Linux

echo
echo "=== Initial image setup ==="
echo

$QEMU_IMG create -f qcow2 $file_name -o size=768M -o zone.size=64M -o \
zone.capacity=64M -o zone.conventional_zones=0 -o zone.max_append_bytes=32M \
-o zone.max_open_zones=6 -o zone.max_active_zones=8 -o zone.mode=host-managed

IMG="--image-opts -n driver=qcow2,file.driver=file,file.filename=$file_name"
QEMU_IO_OPTIONS=$QEMU_IO_OPTIONS_NO_FMT

echo
echo "=== Testing a qcow2 img with zoned format ==="
echo
echo "case 1: test zone operations one by one"

echo "(1) report zones[0]:"
$QEMU_IO $IMG -c "zrp 0 1"
echo
echo "report zones[0~9]:"
$QEMU_IO $IMG -c "zrp 0 10"
echo
echo "report zones[-1]:"  # zones[-1] dictates the last zone
$QEMU_IO $IMG -c "zrp 0x2C000000 2" # 0x2C000000 / 512 = 0x160000
echo
echo
echo "(2) open zones[0], zones[1], zones[-1] then close, finish, reset:"
$QEMU_IO $IMG << EOF
zo 0 0x4000000
zrp 0 1
zo 0x4000000 0x4000000
zrp 0x4000000 1
zo 0x2C000000 0x4000000
zrp 0x2C000000 2
zc 0 0x4000000
zrp 0 1
zc 0x4000000 0x4000000
zrp 0x4000000 1
zc 0x2C000000 0x4000000
zrp 0x2C000000 2
zf 0 0x4000000
zrp 0 1
zf 64M 64M
zrp 0x4000000 2
zf 0x2C000000 0x4000000
zrp 0x2C000000 2
zrs 0 0x4000000
zrp 0 1
zrs 0x4000000 0x4000000
zrp 0x4000000 1
zrs 0x2C000000 0x4000000
zrp 0x2C000000 2
EOF

echo
echo "(3) append write with (4k, 8k) data"
$QEMU_IO $IMG -c "zrp 0 12" # the physical block size of the device is 4096
echo "Append write zones[0], zones[1] twice"
$QEMU_IO $IMG << EOF
zap -p 0 0x1000 0x2000
zrp 0 1
zap -p 0 0x1000 0x2000
zrp 0 1
zap -p 0x4000000 0x1000 0x2000
zrp 0x4000000 1
zap -p 0x4000000 0x1000 0x2000
zrp 0x4000000 1
EOF

echo
echo "Reset all:"
$QEMU_IO $IMG -c "zrp 0 12" -c "zrs 0 768M" -c "zrp 0 12"
echo
echo

echo "case 2: test a sets of ops that works or not"
echo "(1) append write (4k, 4k) and then write to full"
$QEMU_IO $IMG << EOF
zap -p 0 0x1000 0x1000
zrp 0 1
zap -p 0 0x1000 0x1ffd000
zap -p 0 0x1000000 0x1000000
zrp 0 1
EOF

echo "Reset zones[0]:"
$QEMU_IO $IMG -c "zrs 0 64M" -c "zrp 0 1"

echo "(2) write in zones[0], zones[3], zones[8], and then reset all"
$QEMU_IO $IMG << EOF
zap -p 0 0x1000 0x1000
zap -p 0xc000000 0x1000 0x1000
zap -p 0x20000000 0x1000 0x1000
zrp 0 12
zrs 0 768M
zrp 0 12
EOF

echo "case 3: test zone resource management"
echo "(1) write in zones[0], zones[1], zones[2] and then close it"
$QEMU_IO $IMG << EOF
zap -p 0 0x1000 0x1000
zap -p 0x4000000 0x1000 0x1000
zap -p 0x8000000 0x1000 0x1000
zrp 0 12
zc 0 64M
zc 0x4000000 64M
zc 0x8000000 64M
zrp 0 12
EOF

echo "(2) reset all after 3(1)"
$QEMU_IO $IMG << EOF
zrs 0 768M
zrp 0 12
EOF

# success, all done
echo "*** done"
rm -f $seq.full
status=0
