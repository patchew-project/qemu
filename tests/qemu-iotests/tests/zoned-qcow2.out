QA output created by zoned-qcow2

=== Initial image setup ===

Formatting 'zbc.qcow2', fmt=qcow2 cluster_size=65536 extended_l2=off compression_type=zlib zone.mode=host-managed zone.size=67108864 zone.capacity=67108864 zone.conventional_zones=0 zone.max_append_bytes=131072 zone.max_active_zones=12 zone.max_open_zones=10 size=805306368 lazy_refcounts=off refcount_bits=16

=== Testing a qcow2 img with zoned format ===

case 1: test zone operations one by one
(1) report zones[0]:
start: 0x0, len 0x20000, cap 0x20000, wptr 0x0, zcond:1, [type: 2]

report zones[0~9]:
start: 0x0, len 0x20000, cap 0x20000, wptr 0x0, zcond:1, [type: 2]
start: 0x20000, len 0x20000, cap 0x20000, wptr 0x20000, zcond:1, [type: 2]
start: 0x40000, len 0x20000, cap 0x20000, wptr 0x40000, zcond:1, [type: 2]
start: 0x60000, len 0x20000, cap 0x20000, wptr 0x60000, zcond:1, [type: 2]
start: 0x80000, len 0x20000, cap 0x20000, wptr 0x80000, zcond:1, [type: 2]
start: 0xa0000, len 0x20000, cap 0x20000, wptr 0xa0000, zcond:1, [type: 2]
start: 0xc0000, len 0x20000, cap 0x20000, wptr 0xc0000, zcond:1, [type: 2]
start: 0xe0000, len 0x20000, cap 0x20000, wptr 0xe0000, zcond:1, [type: 2]
start: 0x100000, len 0x20000, cap 0x20000, wptr 0x100000, zcond:1, [type: 2]
start: 0x120000, len 0x20000, cap 0x20000, wptr 0x120000, zcond:1, [type: 2]

report zones[-1]:
start: 0x160000, len 0x20000, cap 0x20000, wptr 0x160000, zcond:1, [type: 2]


(2) open zones[0], zones[1], zones[-1] then close, finish, reset:
qemu-io> bad argument count 8 to zo, expected 2 arguments
qemu-io> start: 0x0, len 0x20000, cap 0x20000, wptr 0x0, zcond:1, [type: 2]
qemu-io> qemu-io> start: 0x20000, len 0x20000, cap 0x20000, wptr 0x20000, zcond:3, [type: 2]
qemu-io> qemu-io> start: 0x160000, len 0x20000, cap 0x20000, wptr 0x160000, zcond:3, [type: 2]
qemu-io> qemu-io> start: 0x0, len 0x20000, cap 0x20000, wptr 0x0, zcond:1, [type: 2]
qemu-io> qemu-io> start: 0x160000, len 0x20000, cap 0x20000, wptr 0x160000, zcond:1, [type: 2]
qemu-io> qemu-io> start: 0x0, len 0x20000, cap 0x20000, wptr 0x20000, zcond:14, [type: 2]
qemu-io> qemu-io> start: 0x20000, len 0x20000, cap 0x20000, wptr 0x40000, zcond:14, [type: 2]
start: 0x40000, len 0x20000, cap 0x20000, wptr 0x40000, zcond:1, [type: 2]
qemu-io> qemu-io> start: 0x160000, len 0x20000, cap 0x20000, wptr 0x180000, zcond:14, [type: 2]
qemu-io> qemu-io> start: 0x0, len 0x20000, cap 0x20000, wptr 0x0, zcond:1, [type: 2]
qemu-io> qemu-io> start: 0x20000, len 0x20000, cap 0x20000, wptr 0x20000, zcond:1, [type: 2]
qemu-io> qemu-io> start: 0x160000, len 0x20000, cap 0x20000, wptr 0x160000, zcond:1, [type: 2]
qemu-io> 
(3) append write with (4k, 8k) data
start: 0x0, len 0x20000, cap 0x20000, wptr 0x0, zcond:1, [type: 2]
start: 0x20000, len 0x20000, cap 0x20000, wptr 0x20000, zcond:1, [type: 2]
start: 0x40000, len 0x20000, cap 0x20000, wptr 0x40000, zcond:1, [type: 2]
start: 0x60000, len 0x20000, cap 0x20000, wptr 0x60000, zcond:1, [type: 2]
start: 0x80000, len 0x20000, cap 0x20000, wptr 0x80000, zcond:1, [type: 2]
start: 0xa0000, len 0x20000, cap 0x20000, wptr 0xa0000, zcond:1, [type: 2]
start: 0xc0000, len 0x20000, cap 0x20000, wptr 0xc0000, zcond:1, [type: 2]
start: 0xe0000, len 0x20000, cap 0x20000, wptr 0xe0000, zcond:1, [type: 2]
start: 0x100000, len 0x20000, cap 0x20000, wptr 0x100000, zcond:1, [type: 2]
start: 0x120000, len 0x20000, cap 0x20000, wptr 0x120000, zcond:1, [type: 2]
start: 0x140000, len 0x20000, cap 0x20000, wptr 0x140000, zcond:1, [type: 2]
start: 0x160000, len 0x20000, cap 0x20000, wptr 0x160000, zcond:1, [type: 2]
Append write zones[0], zones[1] twice
qemu-io> After zap done, the append sector is 0x0
qemu-io> start: 0x0, len 0x20000, cap 0x20000, wptr 0x18, zcond:2, [type: 2]
qemu-io> After zap done, the append sector is 0x18
qemu-io> start: 0x0, len 0x20000, cap 0x20000, wptr 0x30, zcond:2, [type: 2]
qemu-io> After zap done, the append sector is 0x20000
qemu-io> start: 0x20000, len 0x20000, cap 0x20000, wptr 0x20018, zcond:2, [type: 2]
qemu-io> After zap done, the append sector is 0x20018
qemu-io> start: 0x20000, len 0x20000, cap 0x20000, wptr 0x20030, zcond:2, [type: 2]
qemu-io> 
Reset all:
start: 0x0, len 0x20000, cap 0x20000, wptr 0x0, zcond:1, [type: 2]
start: 0x20000, len 0x20000, cap 0x20000, wptr 0x20000, zcond:1, [type: 2]
start: 0x40000, len 0x20000, cap 0x20000, wptr 0x40000, zcond:1, [type: 2]
start: 0x60000, len 0x20000, cap 0x20000, wptr 0x60000, zcond:1, [type: 2]
start: 0x80000, len 0x20000, cap 0x20000, wptr 0x80000, zcond:1, [type: 2]
start: 0xa0000, len 0x20000, cap 0x20000, wptr 0xa0000, zcond:1, [type: 2]
start: 0xc0000, len 0x20000, cap 0x20000, wptr 0xc0000, zcond:1, [type: 2]
start: 0xe0000, len 0x20000, cap 0x20000, wptr 0xe0000, zcond:1, [type: 2]
start: 0x100000, len 0x20000, cap 0x20000, wptr 0x100000, zcond:1, [type: 2]
start: 0x120000, len 0x20000, cap 0x20000, wptr 0x120000, zcond:1, [type: 2]
start: 0x140000, len 0x20000, cap 0x20000, wptr 0x140000, zcond:1, [type: 2]
start: 0x160000, len 0x20000, cap 0x20000, wptr 0x160000, zcond:1, [type: 2]


case 2: test a sets of ops that works or not
(1) append write (4k, 4k) and then write to full
qemu-io> bad argument count 8 to zap, expected between 3 and 4 arguments
qemu-io> start: 0x0, len 0x20000, cap 0x20000, wptr 0x0, zcond:1, [type: 2]
qemu-io> zone append failed: Operation not supported
qemu-io> start: 0x0, len 0x20000, cap 0x20000, wptr 0x0, zcond:1, [type: 2]
qemu-io> Reset zones[0]:
start: 0x0, len 0x20000, cap 0x20000, wptr 0x0, zcond:1, [type: 2]
(2) write in zones[0], zones[3], zones[8], and then reset all
qemu-io> After zap done, the append sector is 0x0
qemu-io> After zap done, the append sector is 0x60000
qemu-io> After zap done, the append sector is 0x100000
qemu-io> start: 0x0, len 0x20000, cap 0x20000, wptr 0x10, zcond:2, [type: 2]
start: 0x20000, len 0x20000, cap 0x20000, wptr 0x20000, zcond:1, [type: 2]
start: 0x40000, len 0x20000, cap 0x20000, wptr 0x40000, zcond:1, [type: 2]
start: 0x60000, len 0x20000, cap 0x20000, wptr 0x60010, zcond:2, [type: 2]
start: 0x80000, len 0x20000, cap 0x20000, wptr 0x80000, zcond:1, [type: 2]
start: 0xa0000, len 0x20000, cap 0x20000, wptr 0xa0000, zcond:1, [type: 2]
start: 0xc0000, len 0x20000, cap 0x20000, wptr 0xc0000, zcond:1, [type: 2]
start: 0xe0000, len 0x20000, cap 0x20000, wptr 0xe0000, zcond:1, [type: 2]
start: 0x100000, len 0x20000, cap 0x20000, wptr 0x100010, zcond:2, [type: 2]
start: 0x120000, len 0x20000, cap 0x20000, wptr 0x120000, zcond:1, [type: 2]
start: 0x140000, len 0x20000, cap 0x20000, wptr 0x140000, zcond:1, [type: 2]
start: 0x160000, len 0x20000, cap 0x20000, wptr 0x160000, zcond:1, [type: 2]
qemu-io> qemu-io> start: 0x0, len 0x20000, cap 0x20000, wptr 0x0, zcond:1, [type: 2]
start: 0x20000, len 0x20000, cap 0x20000, wptr 0x20000, zcond:1, [type: 2]
start: 0x40000, len 0x20000, cap 0x20000, wptr 0x40000, zcond:1, [type: 2]
start: 0x60000, len 0x20000, cap 0x20000, wptr 0x60000, zcond:1, [type: 2]
start: 0x80000, len 0x20000, cap 0x20000, wptr 0x80000, zcond:1, [type: 2]
start: 0xa0000, len 0x20000, cap 0x20000, wptr 0xa0000, zcond:1, [type: 2]
start: 0xc0000, len 0x20000, cap 0x20000, wptr 0xc0000, zcond:1, [type: 2]
start: 0xe0000, len 0x20000, cap 0x20000, wptr 0xe0000, zcond:1, [type: 2]
start: 0x100000, len 0x20000, cap 0x20000, wptr 0x100000, zcond:1, [type: 2]
start: 0x120000, len 0x20000, cap 0x20000, wptr 0x120000, zcond:1, [type: 2]
start: 0x140000, len 0x20000, cap 0x20000, wptr 0x140000, zcond:1, [type: 2]
start: 0x160000, len 0x20000, cap 0x20000, wptr 0x160000, zcond:1, [type: 2]
qemu-io> *** done
